clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -384,37 +326,39 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
+                                                            Map<String, List<TopicPartition>> currentAssignment) {
+        if (log.isDebugEnabled()) {
+            log.debug("performing general assign. partitionsPerTopic: {}, subscriptions: {}, currentAssignment: {}",
+                partitionsPerTopic, subscriptions, currentAssignment);
+        }
+
         Map<TopicPartition, ConsumerGenerationPair> prevAssignment = new HashMap<>();
         partitionMovements = new PartitionMovements();
 
-        prepopulateCurrentAssignments(subscriptions, currentAssignment, prevAssignment);
+        prepopulateCurrentAssignments(subscriptions, prevAssignment);
 
-        // a mapping of all topic partitions to all consumers that can be assigned to them
-        final Map<TopicPartition, List<String>> partition2AllPotentialConsumers = new HashMap<>();
-        // a mapping of all consumers to all potential topic partitions that can be assigned to them
-        final Map<String, List<TopicPartition>> consumer2AllPotentialPartitions = new HashMap<>();
+        // a mapping of all topics to all consumers that can be assigned to them
+        final Map<String, List<String>> topic2AllPotentialConsumers = new HashMap<>(partitionsPerTopic.keySet().size());
+        // a mapping of all consumers to all potential topics that can be assigned to them
+        final Map<String, List<String>> consumer2AllPotentialTopics = new HashMap<>(subscriptions.keySet().size());
 
-        // initialize partition2AllPotentialConsumers and consumer2AllPotentialPartitions in the following two for loops
-        for (Entry<String, Integer> entry: partitionsPerTopic.entrySet()) {
-            for (int i = 0; i < entry.getValue(); ++i)
-                partition2AllPotentialConsumers.put(new TopicPartition(entry.getKey(), i), new ArrayList<>());
-        }
+        // initialize topic2AllPotentialConsumers and consumer2AllPotentialTopics
+        partitionsPerTopic.keySet().stream().forEach(
+            topicName -> topic2AllPotentialConsumers.put(topicName, new ArrayList<>()));
 
         for (Entry<String, Subscription> entry: subscriptions.entrySet()) {
             String consumerId = entry.getKey();
-            consumer2AllPotentialPartitions.put(consumerId, new ArrayList<>());
+            List<String> subscribedTopics = new ArrayList<>(entry.getValue().topics().size());
+            consumer2AllPotentialTopics.put(consumerId, subscribedTopics);
             entry.getValue().topics().stream().filter(topic -> partitionsPerTopic.get(topic) != null).forEach(topic -> {
-                for (int i = 0; i < partitionsPerTopic.get(topic); ++i) {
-                    TopicPartition topicPartition = new TopicPartition(topic, i);
-                    consumer2AllPotentialPartitions.get(consumerId).add(topicPartition);
-                    partition2AllPotentialConsumers.get(topicPartition).add(consumerId);
-                }
+                subscribedTopics.add(topic);
```

Comment:
No, it just create a List with the capacity: `topics().size()`. We cannot just create a List with all topics directly, because we need to filter out topics not in `partitionsPerTopic`.